content/blog/posts/2022-01-13-react-objects-and-state.md

```diff
@@ -0,0 +1,171 @@
+---
+authorId: hannakim
+categories:
+  - javascript
+  - react
+date: '2021-11-09'
+slug: react-objects-and-statetitle: 'Handling Objects in React Component State'
+---
+
+## Understanding Primitive and Non-Primitive Data Types
+
+Objects are a useful way to store data when you have a set of related values that you will be accessing or updating together within a given component. To understand how they are used in React, taking a deeper dive into its data type can be helpful.
+
+A [primitive data type](https://developer.mozilla.org/en-US/docs/Glossary/Primitive), like a number or string, holds a value and therefore is inherently immutable.
+
+```
+let greeting = 'hello'
+let hello = 'hello'
+greeting === hello
+
+// Returns true
+// These 2 variables hold the same primitive data as values. If a variable is reassigned, it would take on a different value.
+```
+
+Objects are non-primitive data and have a value as well as an identity. When an object is assigned to a variable, that variable acts as a reference point to their identity. Each object is unique, even if their properties might be identical.
+
+```
+const greeting = { text: 'hello' }
+const hello = { text: 'hello' }
+greeting === hello
+
+// Returns false
+```
+
+While primitives can’t change value without changing reference, objects can change value and **still hold onto its identity**. So two seemingly similar objects are actually two entirely distinct entities to JS.
+
+![The boxes, aka objects, may look the same but are considered unique](/Users/hannakim/Documents/ship-shape/website-nuxt/static/img/blog/cookie-boxes.jpeg)
+
+You can think of an object as a box and its properties as anything you put inside the box. You might put 10 chocolate chip cookies into each box, but they’re still distinct boxes from one another.
+
+## Three Ways to Update Objects with useState
+
+[Data in state should be treated as immutable](https://beta.reactjs.org/learn/updating-objects-in-state) - the values should not be mutated directly, but instead be replaced with a new one. The corresponding `set` method is then called with a new object in order for the change to occur.
+
+This may not seem immediately obvious with objects, since properties of an object *can* technically be updated. However, when you try to handle the change this way with React, it might not update your UI as you’d expect.
+
+```
+const [user, setUser] = useState({ name: '' })
+const handleChange = (e) => {
+   e.preventDefault()
+   user.name = e.target.value
+}
+
+return (
+  <>
```

Comment:
I know nothing about React, so forgive my ignorance, but is `<>` a common thing to use?

Reply:
@rwwagner90 I've been using them when I don't want/need an extra `<div>` in the DOM but need to house 2+ elements

Reply:
@rwwagner90 https://reactjs.org/docs/fragments.html#short-syntax